htlcswitch/circuit_map.go

```diff
@@ -250,6 +255,214 @@ func (cm *circuitMap) initBuckets() error {
 	}, func() {})
 }
 
+// cleanClosedChannels deletes all circuits and keystones related to closed
+// channels. It first reads all the closed channels and caches the ShortChanIDs
+// into a map for fast lookup. Then it iterates the circuit bucket and keystone
+// bucket and deletes items whose ChanID matches the ShortChanID.
+//
+// NOTE: this operation can also be built into restoreMemState since the latter
+// already opens and iterates the two root buckets, circuitAddKey and
+// circuitKeystoneKey. Depending on the size of the buckets, this marginal gain
+// may be worth investigating. Atm, for clarity, this operation is wrapped into
+// its own function.
+func (cm *circuitMap) cleanClosedChannels() error {
+	log.Infof("Cleaning circuits from disk for closed channels")
+
+	// closedChanIDSet stores the short channel IDs for closed channels.
+	closedChanIDSet := make(map[lnwire.ShortChannelID]struct{})
+
+	// circuitKeySet stores the incoming circuit keys of the payment
+	// circuits that need to be deleted.
+	circuitKeySet := make(map[CircuitKey]struct{})
+
+	// keystoneKeySet stores the outgoing keys of the keystones that need
+	// to be deleted.
+	keystoneKeySet := make(map[CircuitKey]struct{})
+
+	// isClosedChannel is a helper closure that returns a bool indicating
+	// the chanID belongs to a closed channel.
+	isClosedChannel := func(chanID lnwire.ShortChannelID) bool {
+		// Skip if the channel ID is zero value. This has the effect
+		// that a zero value incoming or outgoing key will never be
+		// matched and its corresponding circuits or keystones are not
+		// deleted.
+		if chanID.ToUint64() == 0 {
+			return false
+		}
+
+		_, ok := closedChanIDSet[chanID]
+		return ok
+	}
+
+	// Find closed channels and cache their ShortChannelIDs into a map.
+	// This map will be used for looking up relative circuits and keystones.
+	closedChannels, err := cm.cfg.DB.FetchClosedChannels(false)
```

Comment:
Don't consider it a blocker, but since we only need the _short chan ID_ of each of the closed channels, we could add a _new_ method here that just scans the buckets to fetch the keys themselves. This would let us not have to decode the values, and potentially not send the values at all for certain classes of replicated databases.

Reply:
I think the bucket keys are channel outpoints right? I think we'd still load the value bytes and encode the short channel ID. But yeah, this is a good way to optimize the CRUD in our db, and, we could def write more granular methods for specific cases.